samza-test/src/test/java/org/apache/samza/test/table/TestRemoteTableWithBatchEndToEnd.java

```diff
@@ -0,0 +1,250 @@
+    writtenRecords.put(testName, new CopyOnWriteArrayList<>());
+
+    int count = 10;
+    PageView[] pageViews = generatePageViews(count);
+    String profiles = Base64Serializer.serialize(generateProfiles(count));
+
+    int partitionCount = 4;
+    Map<String, String> configs = TestLocalTableEndToEnd.getBaseJobConfig(bootstrapUrl(), zkConnect());
+
+    configs.put("streams.PageView.samza.system", "test");
+    configs.put("streams.PageView.source", Base64Serializer.serialize(pageViews));
+    configs.put("streams.PageView.partitionCount", String.valueOf(partitionCount));
+
+    final RateLimiter readRateLimiter = mock(RateLimiter.class, withSettings().serializable());
+    final RateLimiter writeRateLimiter = mock(RateLimiter.class, withSettings().serializable());
+    final TableRateLimiter.CreditFunction creditFunction = (k, v, args)->1;
+    final StreamApplication app = appDesc -> {
+      RemoteTableDescriptor<Integer, Profile> inputTableDesc = new RemoteTableDescriptor<>("profile-table-1");
+      inputTableDesc
+          .withReadFunction(InMemoryReadFunction.getInMemoryReadFunction(testName, profiles))
+          .withRateLimiter(readRateLimiter, creditFunction, null);
+      if (batchRead) {
+        inputTableDesc.withBatchProvider(new CompactBatchProvider().withmaxBatchDelay(Duration.ofMillis(10)));
+      }
+
+      // dummy reader
+      TableReadFunction readFn = new MyReadFunction();
+
+      RemoteTableDescriptor<Integer, EnrichedPageView> outputTableDesc = new RemoteTableDescriptor<>("enriched-page-view-table-1");
+      outputTableDesc
+          .withReadFunction(readFn)
+          .withWriteFunction(writer)
+          .withRateLimiter(writeRateLimiter, creditFunction, creditFunction);
+      if (batchWrite) {
+        outputTableDesc.withBatchProvider(new CompactBatchProvider().withmaxBatchDelay(Duration.ofMillis(10)));
+      }
+
+      Table<KV<Integer, EnrichedPageView>> outputTable = appDesc.getTable(outputTableDesc);
+
+      Table<KV<Integer, Profile>> inputTable = appDesc.getTable(inputTableDesc);
+
+      DelegatingSystemDescriptor ksd = new DelegatingSystemDescriptor("test");
+      GenericInputDescriptor<PageView> isd = ksd.getInputDescriptor("PageView", new NoOpSerde<>());
+      appDesc.getInputStream(isd)
+          .map(pv -> new KV<>(pv.getMemberId(), pv))
+          .join(inputTable, new PageViewToProfileJoinFunction())
+          .map(m -> new KV(m.getMemberId(), m))
+          .sendTo(outputTable);
+    };
+
+    Config config = new MapConfig(configs);
+    final LocalApplicationRunner runner = new LocalApplicationRunner(app, config);
+    executeRun(runner, config);
+
+    runner.waitForFinish();
+    int numExpected = count * partitionCount;
+    Assert.assertEquals(numExpected, writtenRecords.get(testName).size());
+    Assert.assertTrue(writtenRecords.get(testName).get(0) instanceof EnrichedPageView);
+
+    if (batchRead) {
+      Assert.assertEquals(numExpected, batchReads.get(testName).get());
```

Comment:
Does this mean each batch has got only 1 element in it, can we increase the max duration?